### PR TITLE
UI: Drag and drop creation of sources via .json

### PR DESCRIPTION
**Pull request by Andersama**
Allows .json files to create any type of source, extends the already existing
supported drag and drop feature to handle any supported source.

The .json follows the format that is exported to scene collections. Gives the
user the flexibility of individual transfer of sources between obs-studio as
opposed to whole scene collections.

Known issue: Bug or feature? Scene sources add empty scenes, creating scenes seems to require a few extra special steps, not sure at the moment what that would be. This addition should more than likely be entirely consistent, so scene sources shouldn't add new scenes but instead reference one.

### CODE REVIEW

File: UI/window-basic-main-dropfiles.cpp [hunk 1]
```diff
@@ -105,6 +110,12 @@ void OBSBasic::AddDropSource(const char *data, DropType image)
...
 		break;
+	case DropType_Source:
+		json_settings = obs_data_create_from_json_file(data);
+		settings = obs_data_get_obj(json_settings, "settings");
+		name = obs_data_get_string(json_settings, "name");
+		type = obs_data_get_string(json_settings, "id");
+		break;
 	}
 
 	if (!obs_source_get_display_name(type))
```
[the remark this answers is not in the document]
Comment (Andersama) [reply UI/window-basic-main-dropfiles.cpp]: When I used obs_data_release like this:
```c
	case DropType_Source:
		json_settings = obs_data_create_from_json_file(data);
		settings = obs_data_get_obj(json_settings, "settings");
		name = obs_data_get_string(json_settings, "name");
		type = obs_data_get_string(json_settings, "id");
		obs_data_release(json_settings);
		break;
	}
```
the import would fail because the name value of the string was being released...if this'd cause a memory leak, there's already one in here it seems.

[the remark this answers is not in the document]
Comment (palana) [reply UI/window-basic-main-dropfiles.cpp]: Yes, there already is a memory leak (`settings` leaks at that same point); not asking you to fix that, just pointing out that you shouldn't introduce another memory leak (also, you can either fix this by replacing `obs_data_t json_settings = nullptr;` with `OBSData json_settings;` and adding `obs_data_release(json_settings);` right after `json_settings = obs_data_create_from_json_file(data);`, or you can use something like
```
	if (!obs_source_get_display_name(type)) {
		obs_data_release(json_settings);
		return;
	}
```

[the remark this answers is not in the document]
Comment (Andersama) [reply UI/window-basic-main-dropfiles.cpp]: Thanks for the catch, it didn't cross my mind about the `return;` at that point. I wrapped that part in an if block and properly released both pointers, so this'll solve the memory leak for the other one as well.

File: UI/window-basic-main-dropfiles.cpp [hunk 2]
```diff
@@ -105,10 +110,19 @@ void OBSBasic::AddDropSource(const char *data, DropType image)
...
+		json_settings = obs_data_create_from_json_file(data);
+		settings = obs_data_get_obj(json_settings, "settings");
+		name = obs_data_get_string(json_settings, "name");
+		type = obs_data_get_string(json_settings, "id");
+		break;
 	}
 
-	if (!obs_source_get_display_name(type))
+	if (!obs_source_get_display_name(type)) {
+		obs_data_release(settings);
```
Comment (palana) [inline UI/window-basic-main-dropfiles.cpp]: Please don't add this (`obs_data_release(settings);`) in the same commit; make a separate commit ("UI: Fix drag&drop memory leak" or similar)

Comment (Andersama) [reply UI/window-basic-main-dropfiles.cpp]: That would make more sense to do wouldn't it...

Comment (Andersama) [reply UI/window-basic-main-dropfiles.cpp]: Ah...sorry, don't know why I thought you meant pr / commit as opposed to just commit. Now I'm proper confused.

Comment (palana) [reply UI/window-basic-main-dropfiles.cpp]: Ideally both (separate PR and commit)

Comment (Andersama) [reply UI/window-basic-main-dropfiles.cpp]: There shouldn't be a merge conflict within the if block if I separate these out right?

Comment (palana) [reply UI/window-basic-main-dropfiles.cpp]: There'll likely not be a merge conflict if we follow these steps:
1. you make a separate PR with the leak fix
2. I merge that PR
3. you rebase this PR on top of the fixed code

Comment (Andersama) [reply UI/window-basic-main-dropfiles.cpp]: Alright, well, got step 1. out of the way, I'll rebase after I actually finish my lunch.

Comment (palana) [reply UI/window-basic-main-dropfiles.cpp]: Ready for step 3

Comment (Andersama) [reply UI/window-basic-main-dropfiles.cpp]: Doing a local build just to triple check*, I'll be pushing in maybe 15~
